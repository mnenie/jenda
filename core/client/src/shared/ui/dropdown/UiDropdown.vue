<script setup lang="ts">
import { onClickOutside } from '@vueuse/core'
import { computed, ref, useTemplateRef } from 'vue'

const { direction = 'center' } = defineProps<{
  direction?: 'left' | 'center' | 'right'
}>()

defineSlots<{
  header: () => any
  trigger: () => any
  content: () => any
}>()

const isOpen = ref(false)
const dropdown = useTemplateRef<HTMLElement | null>('dropdown')

function handleDropdown() {
  isOpen.value = !isOpen.value
}
function onDropdownContent() {
  isOpen.value = false
}

onClickOutside(dropdown, () => {
  isOpen.value = false
})

const dropdownStyles = computed(() => {
  if (direction === 'left') {
    return { top: '36px', left: '0px', transform: 'translateX(0%)' }
  }
  else if (direction === 'center') {
    return { top: '36px', left: '50%', transform: 'translateX(-50%)' }
  }
  else {
    return { top: '36px', right: '0px', transform: 'translateX(0%)' }
  }
})
</script>

<template>
  <div ref="dropdown" :class="$style.dropdown">
    <div :class="$style.wrapper" @click="handleDropdown">
      <slot name="trigger" />
    </div>
    <Transition name="dropdown">
      <div v-if="isOpen" :class="$style.inside" :style="dropdownStyles">
        <div v-if="$slots.header" class="text-sm" :class="[$style.header]">
          <slot name="header" />
        </div>
        <div :class="$style.content" @click="onDropdownContent">
          <slot name="content" />
        </div>
      </div>
    </Transition>
  </div>
</template>

<style module lang="scss">
.dropdown {
  position: relative;
  z-index: 9999;

  .wrapper {
    max-height: 29px;
  }

  .inside {
    position: absolute;
    max-height: 24rem;
    height: fit-content;
    min-width: 8rem;
    width: fit-content;
    border-radius: 6px;
    border: 1px solid var(--zinc-200);
    background-color: white;
    color: var(--zinc-900);
    box-shadow: 0 1px 3px 0 rgb(0 0 0 / 0.1), 0 1px 2px -1px rgb(0 0 0 / 0.1);
    text-wrap: nowrap;

    .header {
      width: 100%;
      padding: 0.5rem 10px;
      border-bottom: 1px solid var(--zinc-200);
      display: flex;
      align-items: center;
      justify-content: space-between;
      gap: 20px;
    }

    .content {
      width: 100%;
      padding: 4px;
    }
  }
}

:global(html.dark) {
  .dropdown {
    .inside {
      background-color: rgba(52, 52, 52, 1);
      color: var(--zinc-200);
      border-color: var(--zinc-600);
      .header {
        border-color: var(--zinc-600);
      }
    }
  }
}
</style>
