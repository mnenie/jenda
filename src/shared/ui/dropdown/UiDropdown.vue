<script setup lang="ts">
import { onClickOutside } from '@vueuse/core';
import { ref } from 'vue';

const dropdown = ref<HTMLElement | null>(null);
const isOpen = ref(false);

const handleDropdown = () => {
  isOpen.value = !isOpen.value;
};
const onDropdownContent = () => {
  isOpen.value = false;
};

onClickOutside(dropdown, () => {
  isOpen.value = false;
});

defineSlots<{
  header: () => any;
  trigger: () => any;
  content: () => any;
}>();
</script>

<template>
  <div ref="dropdown" :class="$style.dropdown">
    <div :class="$style.wrapper" @click="handleDropdown">
      <slot name="trigger" />
    </div>
    <Transition name="dropdown">
      <div v-if="isOpen" :class="$style.inside">
        <div v-if="$slots.header" :class="$style.header">
          <slot name="header" />
        </div>
        <div :class="$style.content" @click="onDropdownContent">
          <slot name="content" />
        </div>
      </div>
    </Transition>
  </div>
</template>

<style module lang="scss">
.dropdown {
  position: relative;
  z-index: 9999;

  .wrapper {
    max-height: 29px;
  }

  .inside {
    position: absolute;
    top: 36px;
    right: -4px;
    transform: translateX(0%);
    max-height: 24rem;
    height: fit-content;
    min-width: 8rem;
    width: fit-content;
    border-radius: 6px;
    border: 1px solid var(--zinc-200);
    background-color: white;
    color: var(--zinc-900);
    box-shadow: 0 1px 2px 0 rgb(0 0 0 / 0.05);

    .header {
      width: 100%;
      padding: 0.5rem 10px;
      border-bottom: 1px solid var(--zinc-200);
      display: flex;
      align-items: center;
      justify-content: space-between;
      gap: 20px;
    }

    .content {
      width: 100%;
      padding: 4px;
    }
  }
}

:global(html.dark) {
  .dropdown {
    .inside {
      background-color: rgba(52, 52, 52, 1);
      color: var(--zinc-200);
      border-color: var(--zinc-600);
      .header {
        border-color: var(--zinc-600);
      }
    }
  }
}
</style>
