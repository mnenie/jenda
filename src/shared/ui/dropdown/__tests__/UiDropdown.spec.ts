import { describe, expect, it } from 'vitest';
import { mount } from '@vue/test-utils';
import UiDropdown from '../UiDropdown.vue';

describe('UiDropdown', () => {
  const wrapper = mount(UiDropdown, {
    slots: {
      trigger: '<div>trigger</div>',
      header: '<div>header</div>',
      content: '<div>content</div>'
    }
  });

  it('should render correctly', async () => {
    //@ts-expect-error instance
    wrapper.vm.isOpen = true;
    await wrapper.vm.$nextTick();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('toggles dropdown on trigger', async () => {
    // TODO: think about changing it to $style (vitest.config.ts -> classes())
    const trigger = wrapper.find('[class*="wrapper"]');
    expect(trigger.exists()).toBe(true);
    //@ts-expect-error instance
    wrapper.vm.isOpen = false;

    trigger.trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.find('[class*="inside"]').exists()).toBe(true);

    trigger.trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.find('[class*="inside"]').exists()).toBe(false);
  });

  it('should close dropdown when we clicking outside', () => {
    const _w = mount(UiDropdown, {
      attachTo: document.body,
      slots: {
        trigger: '<div>trigger</div>',
        header: '<div>header</div>',
        content: '<div>content</div>'
      }
    });
    //@ts-expect-error instance
    _w.vm.isOpen = false;

    document.body.click();
    expect(_w.find('.inside').exists()).toBe(false);
  });
});
