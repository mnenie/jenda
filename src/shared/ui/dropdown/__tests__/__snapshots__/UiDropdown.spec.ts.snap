// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UiDropdown > should render correctly 1`] = `
"<div class="_dropdown_e2935a">
  <div class="_wrapper_e2935a">
    <div>trigger</div>
  </div>
  <transition-stub name="dropdown" appear="false" persisted="false" css="true">
    <div class="_inside_e2935a">
      <div class="_header_e2935a">
        <div>header</div>
      </div>
      <div class="_content_e2935a">
        <div>content</div>
      </div>
    </div>
  </transition-stub>
</div>"
`;
